File: server/nestjs/src/base/usecase/product/product.usecase.impl.ts

import { ProductRepository, ProductRepositoryToken } from '../../repository/product.repository';
import { ProductUseCase } from './product.usecase';
import { ProductUIVO, ProductUOVO } from './vos';
import { z } from 'zod';
import { Inject } from '@nestjs/common';

const name = z.string().min(3, 'Name should be at least 3 characters').max(100, 'Name should not exceed 100 characters');
const price = z.number().min(0.5, 'Price should be at least 0.5').max(10000, 'Price should not exceed 10000');
const ProductSchema = z.object({
    name,
    price,
  });

export class ProductUseCaseImpl implements ProductUseCase {
  constructor(@Inject(ProductRepositoryToken)private productRepository: ProductRepository) {}

  async getProduct(id: number): Promise<ProductUOVO> {
    return this.productRepository.getProduct(id);
  }

  async getAllProducts(): Promise<ProductUOVO[]> {
    return this.productRepository.getAllProducts();
  }

  async createProduct(product: ProductUIVO): Promise<ProductUOVO> {
    const validatedProduct = ProductSchema.parse(product);
    return this.productRepository.createProduct(product);
  }

  async updateProduct(id: number, product: ProductUIVO): Promise<ProductUOVO> {
    const validatedProduct = ProductSchema.parse(product);
    return this.productRepository.updateProduct(id, product);
  }

  async deleteProduct(id: number): Promise<void> {
    return this.productRepository.deleteProduct(id);
  }

  async searchProductByNameAndPrice(name?: string, price?: number): Promise<ProductUOVO[]> {
    return this.productRepository.searchProductByNameAndPrice(name, price);
  }
}

File: server/spring-kotlin-mm/app/src/main/kotlin/spring/kotlin/mm/product/usecase/ProductUseCaseImpl.kt
package spring.kotlin.mm.product.usecase
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import spring.kotlin.mm.product.repository.ProductRepository
import spring.kotlin.mm.product.repository.vo.ProductROVO
import spring.kotlin.mm.product.repository.vo.ProductRIVO
import spring.kotlin.mm.product.usecase.vo.ProductUIVO
import spring.kotlin.mm.product.usecase.vo.ProductUOVO

@Service
class ProductUseCaseImpl(@Autowired private val productRepository: ProductRepository) :
    ProductUseCase {
    override fun createProduct(product: ProductUIVO): ProductUOVO {

        val productROVO =  productRepository.save(ProductRIVO(product.id, product.name, product.price))
        return ProductUOVO(productROVO.id, productROVO.name, productROVO.price)
    }

    override fun updateProduct(product: ProductUIVO): ProductUOVO {
        val productROVO =  productRepository.save(ProductRIVO(product.id, product.name, product.price))
        return ProductUOVO(productROVO.id, productROVO.name, productROVO.price)
    }

    override fun deleteProduct(id: Long) {
        productRepository.deleteById(id)
    }

    override fun getProduct(id: Long): ProductUOVO {
        val productROVO = productRepository.findById(id)
        return ProductUOVO(productROVO.id, productROVO.name, productROVO.price)
        // .orElseThrow { throw NotFoundException("Product not found") }
    }

    override fun getProducts(): List<ProductUOVO> {
        return productRepository.findAll()
            .map { ProductUOVO(it.id, it.name, it.price) }
    }

    override fun search(name: String): List<ProductUOVO> {
        return productRepository.findByName(name).map { ProductUOVO(it.id, it.name, it.price) }
    }
}